### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by alvarohuisman**
Entrega del Assignmet 4 (con algunas dudas en el Task 07)

### CODE REVIEW

File: Assignment4/alvarohuisman-22C029/Task06.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  print(s,p,o)
+
+
+# **TASK 6.5: Add UPM as the university where John Smith works**
+
+# In[8]:
+
+
+# TO DO
+g.add((ns.JaneSmith, vcard.Work, Literal('UPM', datatype=XSD.string)))
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task06.py:97]: Work is not defined in vcard.

File: Assignment4/alvarohuisman-22C029/Task07 (1).py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+# In[3]:
+
+
+# TO DO
+from rdflib.plugins.sparql import prepareQuery
+import rdflib
+q1 = prepareQuery('''
+  SELECT ?Subject WHERE { 
+    ?Subject rdfs:subClassOf ?LivingThing.
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task07 (1).py:38]: missing recursive query

File: Assignment4/alvarohuisman-22C029/Task07 (1).py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  SELECT ?Subject WHERE { 
+    ?Subject rdf:type ?Professor.   
+  }
+  ''',
+  initNs = {"rdf": RDF}
+)
+q24 = prepareQuery('''
+  SELECT ?Subject WHERE { 
+    ?Subject rdf:type ?PhDstudent.   
+  }
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task07 (1).py:84]: This should be done in a single query, without having to list all classes by hand!
Use property chains

File: Assignment4/alvarohuisman-22C029/Task07 (1).py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+# In[6]:
+
+
+# TO DO
+from rdflib import FOAF
+
+q4 =  prepareQuery('''
+  SELECT  ?Subject WHERE {
+    ?Subject foaf:knows ?RockySmith.
+    ?Subject rdf:type ?Person.
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task07 (1).py:159]: You are requested to print the name of the persons

File: Assignment4/alvarohuisman-22C029/Task07 (1).py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+)
+
+rocky = rdflib.URIRef("http://somewhere#RockySmith")
+
+# Visualize the results
+for r in g.query(q4, initBindings={'RockySmith':rocky,'Person':person}):
+    print(r.Subject)
+
+
+# **Task 7.5: List the entities who know at least two other entities in the graph**
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task07 (1).py:172]: missing.

And missing all exercises in rdflib

**Review by dgarijo — CHANGES_REQUESTED** [review]
Please have a look at my comments

File: Assignment4/alvarohuisman-22C029/Task_07.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+# %% [markdown]
+# **TASK 7.1: List all subclasses of "LivingThing" with RDFLib and SPARQL**
+
+# %%
+# TO DO
+from rdflib.plugins.sparql import prepareQuery
+import rdflib
+q1 = prepareQuery('''
+  SELECT ?Subject WHERE { 
+    ?Subject rdfs:subClassOf ?LivingThing.
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task_07.py:31]: missing subclasses of subclasses of living thing

File: Assignment4/alvarohuisman-22C029/Task_07.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    ?Subject rdf:type ?Professor.   
+  }
+  ''',
+  initNs = {"rdf": RDF}
+)
+q24 = prepareQuery('''
+  SELECT ?Subject WHERE { 
+    ?Subject rdf:type ?PhDstudent.   
+  }
+  ''',
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task_07.py:75]: Use a single query!

File: Assignment4/alvarohuisman-22C029/Task_07.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+  print(r.Subject, r.FullName, r.Given)
+
+q32 = prepareQuery('''
+  SELECT ?Subject ?Fullname ?Given ?Family WHERE { 
+    ?Subject rdf:type  ?Animal.
+    ?Subject vcard:FN ?FullName.
+    ?Subject vcard:Given ?Given.
+    ?Subject vcard:Family ?Family.
+    
+  }
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task_07.py:125]: Use both in a single query

File: Assignment4/alvarohuisman-22C029/Task_07.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+  GROUP BY ?Subject
+  HAVING (COUNT(?known) >= 2)
+''',
+initNs={"foaf": FOAF})
+
+# Visualize the results
+for r in g.query(q5):
+    print(r.Subject)
+
```
Comment (dgarijo) [inline Assignment4/alvarohuisman-22C029/Task_07.py:177]: Missing exercises in rdflib (7.1 .. 7.5)

**Review by dgarijo — CHANGES_REQUESTED** [review]
Please fix issues and missing exercises

**Comment by alvarohuisman** [comment]
Added all corrections and missing excercises, thank you

**Comment by anaigmo** [comment]
Entregado y corregido, no se admiten más PR

**Pull request merged**